# SDSS SQL Tutorial
---

In the first part of this lab, you were (briefly) introduced to the world of online astronomical databases, and shown how to perform simple searches. The functionality of these databases extends well beyond what you saw in the previous lab. One of the powerful features of these databases is the ability to select out data that fit a certain set of criteria (beyond simply a location on the sky). This can include e.g. selecting all of the bright, well-resolved, nearby galaxies, in order to make your own Hubble classification diagram. Or selecting out all of the stars within a cluster (based on RA, Dec, and distance) in order to create an HR diagram of the cluster, and hence measure its age.

These complex queries are possible, although not with the simple query tools introduced earlier. To perform these tasks you need to learn about the Structured Query Language (SQL) a programming language for performing these queries. To learn SQL, you will work through SDSS SQL tutorial ([http://skyserver.sdss.org/dr14/en/help/howto/search/searchhowtohome.aspx](http://skyserver.sdss.org/dr14/en/help/howto/search/searchhowtohome.aspx)). 

The tutorial can be completed entirely online, but SQL queries can also be completed within python. This notebook demonstrates how to call SQL queries within python. As you work through the online, record and execute your answers in the notebook below. 

### SQL Queries in Python

The next two cells show how to execute the initial SQL query in the tutorial. 

In [1]:
# First, suppress some warnings and import useful packages
import warnings
warnings.filterwarnings('ignore',module='astropy.io.votables.tree')
warnings.filterwarnings('ignore',message='.*unclosed..socket')

import astropy.units as u
from astropy.coordinates import SkyCoord, ICRS
from astroquery.sdss import SDSS #package that allows queries of the SDSS database. 

Basically, the SQL query is recorded in a string variable, and then this string is sent to SDSS using the `query_sql` function.

In [2]:
# First Query

# input the query as a string, and then submit the string to SDSS
query_string='''select ra,dec
from specObj
where ra BETWEEN 140 and 141 AND
dec BETWEEN 20 and 21'''

data = SDSS.query_sql(query_string,verbose=False)

# We can now print the results
print(data)

    ra       dec   
--------- ---------
140.00334 20.774178
140.00636 20.924277
140.01069 20.078161
140.01226 20.538016
140.01318 20.605621
140.01467 20.499697
140.01628 20.849852
140.01957 20.762055
140.02726 20.674341
 140.0331  20.25241
      ...       ...
140.95848 20.611813
 140.9595 20.832956
140.96916 20.290457
140.97144 20.045201
140.97225 20.121144
 140.9787 20.759427
140.97915 20.029572
140.98147 20.817946
140.99308 20.027302
140.99353 20.578518
140.99392 20.423977
Length = 328 rows


The result is an astropy Tables object [(http://docs.astropy.org/en/stable/api/astropy.table.Table.html#astropy.table.Table)](http://docs.astropy.org/en/stable/api/astropy.table.Table.html#astropy.table.Table), which is similar to a dictionary. 

In [3]:
# The column names
print(data.keys())

# Access individual columns
print(data['ra'][:5])

['ra', 'dec']
    ra   
---------
140.00334
140.00636
140.01069
140.01226
140.01318


From here you can complete the rest of the tutorial, recording your answers in the cells below. The examples have also been copied here for your records. While this tutorial does cover a lot of material related to SQL, you only need to go as far as you can before the lab ends (you can save the rest for another time). 

> **Practice 1**: What objects has the SDSS seen in a smaller area of the sky near ra = 140.5, dec = 20.5 (the same area you searched in the previous query)?
>
> Modify the previous query so it will return ra and dec of objects where the ra is between 140.25 and 140.75 and dec is between 20.25 and 20.75. How many objects did the query return?

In [6]:
# Practice 1
query_string=''' select ra, dec
from specObj
where ra between 140.25 and 140.75
and dec between 20.25 and 20.75
'''

data = SDSS.query_sql(query_string,verbose=False)
print(len(data))

89


> **Practice 2**: Which of the objects you found in Practice 1 are galaxies? Modify your query so that it returns the ra, dec, and the best object ID for galaxies (and only galaxies) whose ra is between 194.25 and 194.75 and whose dec is between 2.25 and 2.75.

In [7]:
# Practice 2
query_string=''' select ra, dec, bestObjID
from specObJ
where ra between 194.25 and 194.75
and dec between 2.25 and 2.75
and class='GALAXY'

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

    ra       dec         bestObjID     
--------- --------- -------------------
140.26652  20.27486 1237667430093619286
140.26679 20.410276 1237667293189963907
140.26998 20.665789 1237667430630555834
140.30213 20.516802 1237667293189964312
140.31007 20.374799 1237667293189963927
140.32603 20.627772 1237667430630555857
140.35157 20.735508 1237667430630621359
140.38724 20.602435 1237667430630621652
140.42225 20.443024 1237667293190029773
140.43593 20.450317 1237667293190029434
      ...       ...                 ...
140.70442 20.355897 1237667430093815960
140.71457 20.401091 1237667430093816046
140.72306 20.266393 1237667430093816343
140.72367 20.351012 1237667430093816417
140.72577 20.342547 1237667430093816412
140.72754 20.583188 1237667293190160556
 140.7291 20.400367 1237667430093816489
140.73026 20.441962 1237667430093816530
140.74094 20.320367 1237667430093816437
140.74187  20.34313 1237667430093816467
140.74828 20.343699 1237667430093816484
Length = 40 rows


In [8]:
# Sample Query: Logical Operators
query_string='''select top 10
    z, ra, dec, bestObjID
from
    specObj
where
    class = 'galaxy' 
    and z > 0.3 
    and zWarning = 0
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

    z         ra       dec          bestObjID     
--------- --------- ---------- -------------------
0.3000001 184.86409  52.571314 1237657629529211079
0.3000009 222.98982  42.629408 1237658203986460871
0.3000009 146.39863  5.8334459 1237658423547920434
0.3000012 49.459411   0.847754 1237660241924063461
0.3000018 172.28104  10.234226 1237658493349789816
0.3000027 156.25024  7.6586271 1237658425162858683
0.3000027 231.07321  32.718373 1237662341082382595
0.3000033 197.37712  10.524518 1237662238545478566
0.3000035 184.90501 -3.3095095 1237650760782053596
0.3000037 242.52563  6.0693494 1237662198290186255


In [9]:
#Sample Query: Logical Operators
query_string='''select top 10
    z, ra, dec, bestObjID
from
    specObj
where
    (class = 'galaxy' or class = 'qso')
    and z > 0.3
    and zWarning = 0
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

    z         ra       dec          bestObjID     
--------- --------- ---------- -------------------
0.3000001 184.86409  52.571314 1237657629529211079
0.3000009 222.98982  42.629408 1237658203986460871
0.3000009 146.39863  5.8334459 1237658423547920434
0.3000012 49.459411   0.847754 1237660241924063461
0.3000018 172.28104  10.234226 1237658493349789816
0.3000027 156.25024  7.6586271 1237658425162858683
0.3000027 231.07321  32.718373 1237662341082382595
0.3000033 197.37712  10.524518 1237662238545478566
0.3000035 184.90501 -3.3095095 1237650760782053596
0.3000037 242.52563  6.0693494 1237662198290186255


In [10]:
# Sample Query: Logical Operators
query_string='''select top 10
    ra, dec, modelMag_u, modelMag_g, modelMag_r, modelMag_i, modelMag_z, objID
from
    photoObj
where
    type = 6
    and modelMag_u - modelMag_g < 0.5
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

       ra              dec        modelMag_u ... modelMag_z        objID       
---------------- ---------------- ---------- ... ---------- -------------------
21.7489625734203 20.8928509871811   24.47644 ...   20.93681 1237679541360656765
21.7146801359507 20.4432857186118    23.4531 ...   20.20991 1237679461900353900
22.3833537049599 20.0086760632325   23.31016 ...   21.33505 1237679504864968995
20.4532228010766 19.4146674875702   23.74691 ...    23.3264 1237679503790441055
20.4575029739604 19.4034512992811   22.09286 ...   21.46572 1237679503790441061
23.8528956918851  17.960344971995   25.89019 ...   23.56271 1237679459216982791
23.8678603354314 18.0722111769438     24.668 ...   23.14229 1237679459216982799
24.1930872176298 19.1999599866118    23.1454 ...   21.21239 1237679460827660758
 23.789412470134 18.2858004580844   24.12651 ...   20.76979 1237679502718141036
23.7877842920117 18.2587365072431   24.26909 ...   21.52649 1237679502718141037


> **Practice 3**: What are the reddest galaxies in the area of the sky near ra=141?
> 
> Write a query to search for galaxies between ra = 140.9 and ra=141.1 brighter than g=18.0 for which u-g>2.2. Retrieve the Object ID, ra, dec, and the five final magnitudes.

In [11]:
# Practice 3
query_string='''
select
    ra, dec, objID, u, g, r, i, z
from 
    photoObj
where
    ra between 140.9 and 141.1 and
    g<18.0 and
    u-g>2.2 and
    type=3
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

       ra               dec               objID        ...    i        z    
---------------- ----------------- ------------------- ... -------- --------
140.948230649732   34.483574462363 1237661384381366396 ...   15.211 24.00461
140.948701561789  20.3640889698011 1237667430093946907 ... 13.52119 22.82682
140.948939710061  5.18415672585073 1237658298989477950 ... 15.69538  15.9476
140.948956885199  7.46708493894529 1237660412651241576 ... 15.91807 25.27582
141.024988886179  9.08516071268445 1237660670348689505 ... 10.04595 10.40133
140.950050320666 -2.17559710094289 1237654897379049716 ... 16.77177 16.69215
140.950359995667  2.10954149516854 1237653665254867256 ... 16.53457 17.84776
140.950378430045  2.10952650408391 1237673708259312052 ... 16.94447 27.22122
140.951380330637  -2.0913273185689 1237650804262895728 ... 15.75504 24.33615
141.027608737349  3.46454593595706 1237674460952723503 ... 15.61969 22.83003
             ...               ...                 ... ...      ...      ...

> **Practice 4**: What are the highest-redshift quasars in the SDSS database?
> 
> Write a query to search for quasars for which we have obtained spectra (search the specObj table) with redshifts greater than 4.5 and good measurements (zWarning = 0). Retrieve each quasar's Photo ID, ra dec, and redshift

In [13]:
# Practice 4
query_string='''
select 
    ra, dec, bestObjID, z
from
    specObj
where
    z>4.5 and
    zWarning=0 and
    class = 'qso'

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

    ra        dec           bestObjID         z    
--------- ------------ ------------------- --------
226.87766    55.619792 1237655465388999582 4.500529
166.32833    19.832961 1237668290162131387  4.50241
167.10967   0.61852761 1237674650996769779 4.502814
145.02001    5.4419457 1237674460417754011 4.503839
116.39974    33.178191 1237674289681072697 4.504273
127.76246    52.592732 1237651272960836116 4.504608
 188.6143    29.449435 1237667253480325869 4.504811
34.927581  -0.19360219 1237666407920829375 4.505384
132.93219    23.535822 1237665098459382395 4.505914
 29.34542    2.0443026 1237678618504397427 4.507283
      ...          ...                 ...      ...
359.70154 -0.050893712 1237657190905086176 7.011277
 84.75329    -8.331657                   0 7.011646
86.598473    -8.770931                   0 7.012201
164.41107    25.843305 1237667430103646798  7.01309
 205.4236    20.996424 1237667782306234636 7.014935
359.91718 -0.080629403 1237657190905151776 7.016781
86.339605   

In [14]:
# Sample Query: Joining
query_string='''select top 100
    x.plate, x.mjd,
    s.fiberID,
    p.modelMag_u, p.modelMag_g, p.ModelMag_r, p.ModelMag_i, p.ModelMag_z,
    p.ra, p.dec,
    s.z, p.ObjID
from photoObj p
join specObj s on s.bestobjid = p.objid
join plateX x on x.plateID = s.plateID
where
    s.class = 'qso'
    and s.zwarning = 0
    and s.z between 0.3 and 0.4
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

plate  mjd  fiberID ...        dec             z            ObjID       
----- ----- ------- ... ------------------ --------- -------------------
  435 51882     441 ...   41.7736025629101 0.3626565 1237651191889133881
  266 51630     133 ... -0.790101585801458  0.367843 1237650795146182823
  486 51910     587 ...     61.83930260932 0.3947759 1237651272429994188
  485 51909     217 ...   59.1344403245852 0.3324974 1237651271891419302
  444 51883      17 ...   52.2959788151139 0.3289169 1237651192431640602
  443 51873     185 ...   48.0359554376333  0.352139 1237651190819062130
  441 51868     197 ...   46.1619374339292 0.3525363 1237651190818013682
  474 52000      19 ... -0.718146481855907  0.352988 1237650795144740934
  430 51877     492 ...   14.4605752650757 0.3105766 1237649920581959798
  425 51898     550 ...   14.8252956181591  0.369752 1237649920579862757
  ...   ...     ... ...                ...       ...                 ...
  616 52442     470 ...   54.2451736767219 0.366227

> **Practice 5**: How can you look up image data, plates, and spectra of moderately bright galaxies?
> 
> Write a query to find 100 galaxies for which we have spectra that have g magnitude between 17 and 17.4 and redshift less than 0.05. For each galaxy retrieve the object ID, the five magnitudes, the redshift, the plate/MJD number, and the fiber number.

In [16]:
# Practice 5
query_string='''
select top 100
    s.bestObjID, p.u, p.g, p.r, p.i, p.z, s.fiberID, s.plate, s.mjd
from specObJ s
join photoObJ p on s.bestObjID = p.ObjID
where
    s.z <0.05 and
    p.g between 17 and 17.4 and
    s.class='galaxy' and
    s.zwarning = 0
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

     bestObjID         u        g        r     ...    z     fiberID plate  mjd 
------------------- -------- -------- -------- ... -------- ------- ----- -----
1237648720685301897 18.27173  17.0717 16.65145 ... 16.27766     144   274 51913
1237648721752555640 19.13766 17.23995 16.29899 ... 15.38905     383   266 51630
1237648721752621240 18.38335 17.32703 17.39917 ... 17.59595     438   266 51630
1237648720148693392 18.86752 17.33773 16.80683 ... 16.38996     245   275 51910
1237648721219485979 18.58894 17.37458 16.94893 ... 16.77008     234   271 51883
1237648704577601678 18.48543   17.339 16.90179 ... 16.48684     637   287 52023
1237649918968660092 19.32904 17.20216 16.65694 ... 16.56159       8   424 51893
1237648720685826207 19.03796 17.35102 16.56406 ... 15.90696     188   275 51910
1237648721753276559 18.36063 17.22968 16.69767 ... 16.30265     640   266 51630
1237648720149283041 18.41814 17.17549 16.63656 ...  16.1021     282   276 51909
                ...      ...      ...   

---
You can **stop here**, or continue on if you have time.

---

In [ ]:
# Sample Query: Aggregate functions
query_string='''select 
    min(dec) as min_dec, max(dec) as max_dec, avg(dec) as avg_dec
from
    photoObj
where
    run = 5112
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

In [ ]:
# Sample Query: The Group By command
query_string='''select 
    class, count(z) as num_redshift
from 
    specObj
where
    z between 0.5 and 1
group by
    class
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

In [ ]:
# Sample Query: The Order By command
query_string='''select mjd,plate
from
    plateX
where
    plate <= 275
order by mjd
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

> **Practice 7**: What are the northernmost and southernmost objects with spectra measured by the SDSS?

In [ ]:
# Practice 7
query_string='''

'''

data = SDSS.query_sql(query_string,verbose=False)

# We can now print the results
print(data)

> **Practice 8**: What is the redshift of the nearest galaxy whose spectrum was measured by the SDSS with high confidence (zWarning=0)? 
>
> Compare the distance you found to the distance to the Andromeda Galaxy (2 million light-years) and the Whirlpool Galaxy (37 millions light-years). Does the distance you found seem reasonable?

In [ ]:
# Practice 8
query_string='''

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data['redshift'][0],data['redshift'][0]/(7.11e-11))

> **Practice 9**: What field has galaxies with the highest average redshifts in run=5112, camcol=1?
>
> Be sure you are searching fields (as run-camcol-field) for galaxies. Also look at how many spectrally measured galaxies are in the field - make sure you don't pick a field with only one or two galaxies! Also note that this query will probably take a long time to execute.

In [ ]:
# Practice 9
query_string='''

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

---
You can **stop here**, or continue on if you have time.

---

In [ ]:
# Sample Query: Functions
# Search for all objects within 5 arcminutes of ra=140, dec = 20
query_string='''SELECT
    p.ObjID, p.ra, p.dec, p.u, p.g, p.r, p.i, p.z
FROM photoObj p
JOIN dbo.fGetNearbyObjEq(140,20,5) n ON n.objID = p.objID
WHERE
    p.type = 3
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

In [ ]:
# Sample Query: Flags
# Find English names for flags of all stars around the point 175,1
query_string='''SELECT
    p.ObjID, p.flags, dbo.fPhotoFlagsN(p.flags)
FROM photoObj p
JOIN dbo.fGetNearbyObjEq(140,20,5) n ON n.objID = p.objID
WHERE
    p.type = 6
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

In [ ]:
# Sample Query: Flags
query_string='''SELECT
    p.ObjID, p.ra, p.dec, dbo.fPhotoFlagsN(p.flags)
FROM photoObj p
JOIN dbo.fGetNearbyObjEq(140,20,5) n ON n.objID = p.objID
WHERE
    (p.flags & dbo.fPhotoFlags('SATURATED')) = 0
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

> **Practice 10**: In the field 5112-6-119, what percentage of all objects detected by SDSS are too close to the edge of their fields to be trusted?
>
> Hint: Use two searches, one with a flag and one without. Search run=5112, camcol=6, field=119

In [ ]:
#Practice 10
query_string='''

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

In [ ]:
#Practice 10
query_string='''

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

> **Practice 11**: Choose a galaxy cluster from SkyServer's Famous Places tool. Write a query to select galaxies in the cluster, and only galaxies in the cluster
>
> Hint: After you pick a cluster, use the Navigation Tool to examine the cluster. Guess which galaxies belong to the cluster - you should be able to tell just by looking. Click on 5-10 galaxies and save them in your online notebook. Open the notebook to look for features that the cluster galaxies have in common. Guess the center position and radius of the galaxies. Then, write a query that uses what you have learned to search for the cluster galaxies.

In [ ]:
# Practice 11
query_string='''

'''

data = SDSS.query_sql(query_string,verbose=False)

# We can now print the results
print(data)

- Challenge 1: What percentage of galaxies have spectral redshifts measured? What percentage have photometric redshifts taken? What are the advantages of using photometric redshifts? Try to compare photometric and spectral redshifts; how accurate are photometric redshifts?
- Challenge 2: What are the limits in ra and dec of stripes 42 and 43, two of the SDSS's diagonal stripes?
- Challenge 3: Look at colors and spectra of stars, and find stars consistent with white dwarfs. Create a list of white dwarfs in the SDSS database.
- Challenge 4: What are the largest galaxies in the SDSS, in terms of size? Hint: Look at surface brightness and ellipticity.
- Challenge 5: Find all objects with spectra classified as unknown.
- Challenge 6: Find the broad absorption line (BAL) quasars in the SDSS database. At what redshift are most BAL quasars found?
- Challenge 7: Find variable stars in the SDSS (stars imaged more than once whose magnitude changed by more than 0.1 between observations). How variable are the stars?

SQL is a common language for querying databases. The International Virtual Observatory Association supports use of the Astronomy Database Query Language (ADQL), and is used by e.g. GAIA [https://www.gaia.ac.uk/data/gaia-data-release-1/adql-cookbook](https://www.gaia.ac.uk/data/gaia-data-release-1/adql-cookbook). Sincd ADQL is built on from SQL, much of the syntax is the same. 